### PR TITLE
New rewrite rule for replacing parens with braces in for expressions

### PR DESCRIPTION
**Pull request by caoilte**
- Only applied on expressions with more than one generator
- Also replaces semi-colons with newlines
- Works on ForYield and For

As suggested in https://github.com/olafurpg/scalafmt/issues/470

### CODE REVIEW

**Comment by olafurpg** [comment]
Here's the diff from running the rewrite on Term.For:
https://github.com/olafurpg/scala-repos/pull/9/commits/d3f8da790b51f8fc34f4bae5d2c863bd3d3d885e

I'm OK with rewriting Term.For by default. If someone would like more control we can later introduce a flag.

File: core/src/main/scala/org/scalafmt/rewrite/ReplaceForParensWithBraces.scala [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    builder.result()
+  }
+
+  override def rewrite(code: Tree, ctx: RewriteCtx): Seq[Patch] = {
+    val builder = Seq.newBuilder[Patch]
+
+    code.collect {
+      case fy: Term.ForYield if fy.enums.size > 1 =>
```
Comment (olafurpg) [inline core/src/main/scala/org/scalafmt/rewrite/ReplaceForParensWithBraces.scala]: How about `fy.enums.count(_.is[Enumerator.Generator]) > 1`?

I think this case should maybe not be rewritten.
```diff
-    for (sup <- direct if !res.contains(sup)) res ++= sup.getSupers
+    for {
+      sup <- direct if !res.contains(sup)
+    } res ++= sup.getSupers
```

Comment (caoilte) [inline core/src/main/scala/org/scalafmt/rewrite/ReplaceForParensWithBraces.scala]: Oops! That was me misunderstanding the shape of the ADT! Will put that fix in with a test.

File: core/src/main/scala/org/scalafmt/rewrite/ReplaceForParensWithBraces.scala [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+  *   } yield (a, b)
+  *
+  */
+object ReplaceForParensWithBraces extends Rewrite {
+
+  def findForParens(forTokens: Tokens, ctx: RewriteCtx): Option[(Token, Token)] = {
+    import ctx.tokenTraverser._
+
+    @tailrec
+    def nextValidLeftParen(token: Token):Option[Token] = {
```
Comment (caoilte) [inline core/src/main/scala/org/scalafmt/rewrite/ReplaceForParensWithBraces.scala]: That's a great suggestion. The only change I'd like to make is on the return type. I think `Option[Token]` would be more idiomatic.

``` scala
def find(start: Token)(f: Token => Boolean): Option[Token]
def reverseFind(start: Token)(f: Token => Boolean): Option[Token]
```

File: core/src/main/scala/org/scalafmt/rewrite/ReplaceForParensWithBraces.scala [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+  *
+  * becomes,
+  *
+  *   for {
+  *     a <- as
+  *     b <- bs if b > 2
+  *   } yield (a, b)
+  *
+  */
+object ReplaceForParensWithBraces extends Rewrite {
```
Comment (caoilte) [inline core/src/main/scala/org/scalafmt/rewrite/ReplaceForParensWithBraces.scala]: I very much agree!

File: core/src/main/scala/org/scalafmt/rewrite/ReplaceForParensWithBraces.scala (hunk 2 above)
Comment (olafurpg) [inline core/src/main/scala/org/scalafmt/rewrite/ReplaceForParensWithBraces.scala]: Right, of course, I meant `Token`!

File: core/src/main/scala/org/scalafmt/rewrite/PreferCurlyFors.scala [hunk 4]
```diff
@@ -22,42 +21,29 @@ import scala.meta.{Tree, _}
...
-
-    forEnumerators.flatMap { enumerator =>
-      enumerator.tokens.headOption.flatMap(prevSemiColon).toList
-    }
+    for {
+      enumerator <- forEnumerators
+      token <- enumerator.tokens.headOption.toIterable
+      semicolon <- reverseFind(token)(_.isNot[Whitespace]).filter(_.is[Token.Semicolon]).toIterable
+    } yield {
+      println(semicolon)
```
Comment (olafurpg) [inline core/src/main/scala/org/scalafmt/rewrite/PreferCurlyFors.scala]: Remove

File: core/src/test/resources/rewrite/PreferCurlyForYields.stat [hunk 5]
```diff
@@ -26,7 +26,19 @@ object a {
     for (a <- as) yield a
   }
 }
-<<< replaces parens with braces and removes semi-colons on multi line multi generator for yield expression
+<<< doesn't change single generator for-yield expression with guard that uses parens
+object a {
+  def x(as: List[Int]): List[Int] = {
```
Comment (olafurpg) [inline core/src/test/resources/rewrite/PreferCurlyForYields.stat]: Remove

File: core/src/test/resources/rewrite/PreferCurlyFors.stat [hunk 6]
```diff
@@ -34,6 +34,18 @@ object a {
     }
   }
 }
+<<< doesn't change single generator for expression with guard that uses parens
+object a {
+  def x(as: List[Int]): List[Int] = {
```
Comment (olafurpg) [inline core/src/test/resources/rewrite/PreferCurlyFors.stat]: Remove

File: core/src/test/resources/rewrite/PreferCurlyFors.stat [hunk 7]
```diff
@@ -34,6 +34,18 @@ object a {
...
+    for (a <- as if a < 5) yield a
+  }
+}
+>>>
+object a {
+  def x(as: List[Int]): List[Int] = {
```
Comment (olafurpg) [inline core/src/test/resources/rewrite/PreferCurlyFors.stat]: Remove

File: core/src/main/scala/org/scalafmt/rewrite/PreferCurlyFors.scala (hunk 4 above)
Comment (olafurpg) [inline core/src/main/scala/org/scalafmt/rewrite/PreferCurlyFors.scala]: Remove println + redundant enclosing curly braces.

File: core/src/test/resources/rewrite/PreferCurlyForYields.stat [hunk 8]
```diff
@@ -0,0 +1,74 @@
+rewrite.rules = [PreferCurlyFors]
+<<< replaces parens with braces and removes semi-colons on single line multi generator for-yield expression
+object a {
+  def x(as: List[Int], bs: List[Int]): List[(Int, Int)] = {
```
Comment (olafurpg) [inline core/src/test/resources/rewrite/PreferCurlyForYields.stat]: Could you remove this def in all of the tests? I think it's enough to only test

```scala
object a {
  for (...) yield x
}
```

**Comment by olafurpg** [comment]
I'm slightly confused by this new review system. There's only two comments on removing defs and a println statement. Then I believe we can go ahead and merge :D

File: core/src/test/resources/rewrite/PreferCurlyForYields.stat (hunk 8 above)
Comment (caoilte) [inline core/src/test/resources/rewrite/PreferCurlyForYields.stat]: Will do

File: core/src/main/scala/org/scalafmt/rewrite/PreferCurlyFors.scala (hunk 4 above)
Comment (caoilte) [inline core/src/main/scala/org/scalafmt/rewrite/PreferCurlyFors.scala]: *face palm*

**Comment by caoilte** [comment]
Pushed. I think.

**Comment by olafurpg** [comment]
Excellent, thank you for putting this together!

I compared the output from the new version with the older ReplaceCurlyXXX implemenatation and the only diffs were related to for comprehensions with a single generator and an `if` guard. We may later choose to add a configuration option that rewrites those as well. However, I prefer we start conservatively.

LGTM 🎉

**Comment by caoilte** [comment]
Woo! That was fun. Thanks for guiding me through it. I'll think I'll look for another issue to work on this week!

**Comment by olafurpg** [comment]
>  I'll think I'll look for another issue to work on this week

Awesome! Don't hesitate to ask me if there's anything.

**Pull request merged**